addons/product/models/product_product.py

```diff
@@ -533,50 +533,70 @@ def get_display_name(name, code):
+                or self.search_fetch(expression.AND([domain, [('barcode', '=', name)]]), ['display_name'], limit=limit)
+        if not products and is_positive:
+            # Do not merge the 2 next lines into one single search, SQL search performance would be abysmal
+            # on a database with thousands of matching products, due to the huge merge+unique needed for the
+            # OR operator (and given the fact that the 'name' lookup results come from the ir.translation table
+            # Performing a quick memory merge of ids in Python will give much better performance
+            products = self.search_fetch(expression.AND([domain, [('default_code', operator, name)]]), ['display_name'], limit=limit)
+            limit_rest = limit and limit - len(products)
+            if limit_rest is None or limit_rest > 0:
+                products |= self.search_fetch(expression.AND([domain, [('id', 'not in', products.ids)], [('name', operator, name)]]), ['display_name'], limit=limit_rest)
+        elif not products:
+            assert not is_positive, 'operator is negative'
+            domain_neg = [
+                ('name', operator, name),
+                '|', ('default_code', operator, name), ('default_code', '=', False),
+            ]
+            products = self.search_fetch(expression.AND([domain, domain_neg]), ['display_name'], limit=limit)
+        if not products and operator in positive_operators and (m := re.search(r'(\[(.*?)\])', name)):
+            match_domain = [('default_code', '=', m.group(2))]
+            products = self.search_fetch(expression.AND([domain, match_domain]), ['display_name'], limit=limit)
+        if not products and (partner_id := self.env.context.get('partner_id')):
             # still no results, partner in context: search on supplier info as last hope to find something
-            if not product_ids and self._context.get('partner_id'):
-                suppliers_ids = self.env['product.supplierinfo']._search([
-                    ('partner_id', '=', self._context.get('partner_id')),
-                    '|',
-                    ('product_code', operator, name),
-                    ('product_name', operator, name)])
-                if suppliers_ids:
-                    product_ids = self._search([('product_tmpl_id.seller_ids', 'in', suppliers_ids)], limit=limit, order=order)
-        else:
-            product_ids = self._search(domain, limit=limit, order=order)
-        return product_ids
+            supplier_domain = [
+                ('partner_id', '=', partner_id),
+                '|',
+                ('product_code', operator, name),
+                ('product_name', operator, name),
+            ]
+            match_domain = [('product_tmpl_id.seller_ids', 'any', supplier_domain)]
+            products = self.search_fetch(expression.AND([domain, match_domain]), ['display_name'], limit=limit)
+        return [(product.id, product.display_name) for product in products]
```

Comment:
sudo to compute the display_name maybe ?

Reply:
If you can't access `display_name`, it wouldn't be returned by `search_fetch()`, wouldn't it?